internal/resource/client_service_test.go

```diff
@@ -402,6 +402,127 @@ var _ = Context("ClientServices", func() {
+				}))
+			})
+
+			It("overrides ServiceSpec", func() {
+				var IPv4 corev1.IPFamily = "IPv4"
+				ten := int32(10)
+				instance.Spec.Override.ClientService = &rabbitmqv1beta1.ClientService{
+					Spec: &corev1.ServiceSpec{
+						Ports: []corev1.ServicePort{
+							{
+								Protocol: corev1.ProtocolUDP,
+								Port:     12345,
+								Name:     "my-new-port",
+							},
+						},
+						Selector: map[string]string{
+							"a-selector": "a-label",
+						},
+						Type:                     "NodePort",
+						SessionAffinity:          "ClientIP",
+						LoadBalancerSourceRanges: []string{"1000", "30000"},
+						ExternalName:             "my-external-name",
+						ExternalTrafficPolicy:    corev1.ServiceExternalTrafficPolicyTypeLocal,
+						HealthCheckNodePort:      1234,
+						PublishNotReadyAddresses: false,
+						SessionAffinityConfig: &corev1.SessionAffinityConfig{
+							ClientIP: &corev1.ClientIPConfig{
+								TimeoutSeconds: &ten,
+							},
+						},
+						IPFamily:     &IPv4,
+						TopologyKeys: []string{"a-topology-key"},
+					},
+				}
+
+				err := serviceBuilder.Update(svc)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(svc.Spec.Ports).To(ConsistOf(
+					corev1.ServicePort{
+						Name:     "amqp",
+						Port:     5672,
+						Protocol: corev1.ProtocolTCP,
+					},
+					corev1.ServicePort{
+						Name:     "management",
+						Port:     15672,
+						Protocol: corev1.ProtocolTCP,
+					},
+					corev1.ServicePort{
+						Protocol: corev1.ProtocolUDP,
+						Port:     12345,
+						Name:     "my-new-port",
+					},
+				))
+				Expect(svc.Spec.Selector).To(Equal(map[string]string{"a-selector": "a-label", "app.kubernetes.io/name": "foo"}))
+				Expect(svc.Spec.Type).To(Equal(corev1.ServiceTypeNodePort))
+				Expect(svc.Spec.SessionAffinity).To(Equal(corev1.ServiceAffinityClientIP))
+				Expect(svc.Spec.LoadBalancerSourceRanges).To(Equal([]string{"1000", "30000"}))
+				Expect(svc.Spec.ExternalName).To(Equal("my-external-name"))
+				Expect(svc.Spec.ExternalTrafficPolicy).To(Equal(corev1.ServiceExternalTrafficPolicyTypeLocal))
+				Expect(svc.Spec.HealthCheckNodePort).To(Equal(int32(1234)))
+				Expect(svc.Spec.PublishNotReadyAddresses).To(BeFalse())
+				Expect(*svc.Spec.SessionAffinityConfig.ClientIP.TimeoutSeconds).To(Equal(int32(10)))
+				Expect(*svc.Spec.IPFamily).To(Equal(corev1.IPv4Protocol))
+				Expect(svc.Spec.TopologyKeys).To(Equal([]string{"a-topology-key"}))
+			})
+
+			It("ensures override takes precedence when same property is set both at the top level and at the override level", func() {
+				instance.Spec.Service.Type = "LoadBalancer"
+				instance.Spec.Override.ClientService = &rabbitmqv1beta1.ClientService{
+					Spec: &corev1.ServiceSpec{
+						Type: corev1.ServiceTypeNodePort,
+					},
+				}
+
+				err := serviceBuilder.Update(svc)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(svc.Spec.Type).To(Equal(corev1.ServiceTypeNodePort))
+			})
```

Comment:
Would this test cover the case when we try to override a default port e.g. amqp? Or should we add a specific test for that?

Reply:
Any default port (amqp, empd, and management that the operator set) cannot be overwritten by patching because the svc.spec.ports is a list with a patch merge key of `port`. Override can only give the port a new name or a new target port, but not delete it. I'm not sure about the value of testing that behavior here.

Reply:
That makes complete sense, no need to test that case. We should document this tho in our docs, as it's reasonable to think that the merge key is the port name.